airbyte-cdk/bulk/toolkits/load-s3/src/main/kotlin/io/airbyte/cdk/load/command/s3/S3Client.kt

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.cdk.load.command.s3
+
+import aws.sdk.kotlin.runtime.auth.credentials.StaticCredentialsProvider
+import aws.sdk.kotlin.services.s3.model.ListObjectsRequest
+import io.airbyte.cdk.load.command.aws.AWSAccessKeyConfigurationProvider
+import io.airbyte.cdk.load.file.ObjectStorageClient
+import io.airbyte.cdk.load.message.RemoteObject
+import io.micronaut.context.annotation.Factory
+import io.micronaut.context.annotation.Secondary
+import jakarta.inject.Singleton
+
+class S3Object(
+    override val key: String,
+) : RemoteObject()
+
+class S3Client(
+    private val client: aws.sdk.kotlin.services.s3.S3Client,
+    private val bucketConfig: S3BucketConfiguration
+) : ObjectStorageClient<S3Object> {
+    override suspend fun list(prefix: String): List<S3Object> {
+        val request = ListObjectsRequest {
+            bucket = bucketConfig.s3BucketName
+            this.prefix = prefix
+        }
+        return client.listObjects(request).contents?.mapNotNull {
+            it.key?.let { key -> S3Object(key) }
+        }
+            ?: emptyList()
+    }
+}
+
+@Factory
+class S3ClientFactory(
+    private val keyConfig: AWSAccessKeyConfigurationProvider,
+    private val bucketConfig: S3BucketConfigurationProvider
+) {
+
+    @Singleton
+    @Secondary
+    fun make(): S3Client {
+        val credentials = StaticCredentialsProvider {
```

Comment:
is this really the recommended way? (our current s3 code uses `BasicAWSCredentials` + `AWSStaticCredentialsProvider`, but idk if you're using a different sdk) - https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/java/airbyte-cdk/s3-destinations/src/main/kotlin/io/airbyte/cdk/integrations/destination/s3/credential/S3AccessKeyCredentialConfig.kt#L18

Reply:
I upgraded to the kotlin sdk because A) our existing code threw a deprecation warning B) it provides actual non-blocking suspend funs.